flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_reader.go

```diff
@@ -3,75 +3,243 @@ package ioutils
+	metadata, err := s.store.Head(ctx, s.errorFilePath)
 	if err != nil {
-		if storage.IsNotFound(err) {
-			return io.ExecutionError{
-				IsRecoverable: true,
-				ExecutionError: &core.ExecutionError{
-					Code:    "ErrorFileNotFound",
-					Message: err.Error(),
-					Kind:    core.ExecutionError_SYSTEM,
-				},
-			}, nil
-		}
-		return io.ExecutionError{}, errors.Wrapf(err, "failed to read error data from task @[%s]", r.outPath.GetErrorPath())
+		return false, errors.Wrapf(err, "failed to read error file @[%s]", s.errorFilePath)
+	}
+	err = s.validatePayloadSize(metadata)
+	if err != nil {
+		return false, errors.Wrapf(err, "error file @[%s] failed payload size validation", s.errorFilePath)
 	}
+	return metadata.Exists(), nil
+}
 
+func errorDoc2ExecutionError(errorDoc *core.ErrorDocument, errorFilePath storage.DataReference) io.ExecutionError {
 	if errorDoc.Error == nil {
 		return io.ExecutionError{
 			IsRecoverable: true,
 			ExecutionError: &core.ExecutionError{
 				Code:    "ErrorFileBadFormat",
-				Message: fmt.Sprintf("error not formatted correctly, nil error @path [%s]", r.outPath.GetErrorPath()),
+				Message: fmt.Sprintf("error not formatted correctly, nil error @path [%s]", errorFilePath),
 				Kind:    core.ExecutionError_SYSTEM,
 			},
-		}, nil
+		}
 	}
-
-	ee := io.ExecutionError{
+	executionError := io.ExecutionError{
 		ExecutionError: &core.ExecutionError{
 			Code:    errorDoc.Error.Code,
 			Message: errorDoc.Error.Message,
 			Kind:    errorDoc.Error.Origin,
+			Worker:  errorDoc.Error.Worker,
 		},
 	}
 
 	if errorDoc.Error.Kind == core.ContainerError_RECOVERABLE {
-		ee.IsRecoverable = true
+		executionError.IsRecoverable = true
+	}
+
+	return executionError
+}
+
+func (s *SingleFileErrorRetriever) GetError(ctx context.Context) (io.ExecutionError, error) {
+	errorDoc := &core.ErrorDocument{}
+	err := s.store.ReadProtobuf(ctx, s.errorFilePath, errorDoc)
+	if err != nil {
+		if storage.IsNotFound(err) {
+			return io.ExecutionError{
+				IsRecoverable: true,
+				ExecutionError: &core.ExecutionError{
+					Code:    "ErrorFileNotFound",
+					Message: err.Error(),
+					Kind:    core.ExecutionError_SYSTEM,
+				},
+			}, nil
+		}
+		return io.ExecutionError{}, errors.Wrapf(err, "failed to read error data from task @[%s]", s.errorFilePath)
+	}
+
+	return errorDoc2ExecutionError(errorDoc, s.errorFilePath), nil
+}
+
+type EarliestFileErrorRetriever struct {
+	baseErrorRetriever
+	errorDirPath           storage.DataReference
+	canonicalErrorFilename string
+}
+
+func (e *EarliestFileErrorRetriever) parseErrorFilename() (errorFilePathPrefix storage.DataReference, errorFileExtension string, err error) {
+	// If the canonical error file name is error.pb, we expect multiple error files
+	// to have name error<suffix>.pb
+	pieces := strings.Split(e.canonicalErrorFilename, ".")
+	if len(pieces) != 2 {
+		err = errors.Errorf("expected canonical error filename to have a single dot (.), got %d", len(pieces))
+		return
+	}
```

Comment:
We should probably do this once in the constructor and store the pieces in the struct... they will never change after

Reply:
Addressed. The reason I have not done this in the first place was that, it impacts many call sites as the constructor would have to return an error. Had to bite the bullet and just make that larger change. Please review.